File: leetcode_2140/SolvingQuestionsWithBrainpower.py
class SolvingQuestionsWithBrainpower:
    def mostPoints(self, questions: list[list[int]]) -> int:
        n = len(questions)
        cache = [0] * (n)
        
        # Bottom-up DP approach
        for i in reversed(range(n)):
            points, brainpowers = questions[i]
            next_index = i + 1 + brainpowers 
            choose = points + (cache[next_index] if next_index < n else 0)
            skip = (cache[i + 1] if i + 1 < n else 0)
            cache[i] = max(choose, skip)

        return cache[0]
        
        # Memoization approach
        # def backtracking(i):
        #     if i >= n: return 0
        #     if cache[i]: return cache[i]
            
        #     points, brainpowers = questions[i]
        #     cache[i] = max(
        #         backtracking(i + 1), # skip the current question
        #         points + backtracking(i + 1 + brainpowers) # solve the current question
        #     )
        #     return cache[i]
        # return backtracking(0)
    
def main():
    questions = [[3, 2], [4, 3], [4, 4], [2, 5]]
    solver = SolvingQuestionsWithBrainpower()
    result = solver.mostPoints(questions)
    print(result)  # Output: 5
    
if __name__ == "__main__":
    main()

File: leetcode_2140/SolvingQuestionsWithBrainpower.java
package leetcode_2140;

import java.util.Arrays;

class SolvingQuestionsWithBrainpower{
    public long mostPoints(int[][] questions) {
        int n = questions.length;
        long[] cache = new long[questions.length];
        Arrays.fill(cache, 0);
        
        // Memoization Approach
        // return backtracking(questions, cache, 0);

        // Bottom Up Approach / Tabulation Approach
        for(int i = n - 1; i >= 0; i--) {
            
            long points = questions[i][0];
            int brainpowers = questions[i][1];
            int nextIndex = i + 1 + brainpowers;
            // If we choose the current question
            long chooseCurrent = points + (nextIndex < n ? cache[nextIndex] : 0);
            // If we skip the current question
            long skipCurrent = (i + 1 < n ? cache[i + 1] : 0);

            cache[i] = Math.max(chooseCurrent, skipCurrent);
        }

        return cache[0];
    }

    private long backtracking(int[][] questions, long[] cache, int index) {
        if(index >= questions.length) return 0;
        if(cache[index] != 0)
            return cache[index];
        
        long points = questions[index][0];
        int brainpowers = questions[index][1];

        cache[index] = Math.max(
            backtracking(questions, cache, index + 1), // skip the current question
            points + backtracking(questions, cache, index + 1 + brainpowers) // choose the current question
        );

        return cache[index];
    }

    public static void main(String[] args) {
        SolvingQuestionsWithBrainpower obj = new SolvingQuestionsWithBrainpower();
        int[][] questions = {
            {3, 2},
            {4, 3},
            {4, 4},
            {2, 5}
        };
        long result = obj.mostPoints(questions);
        System.out.println(result); 
    }
}